### PR TITLE
feat (provider/openai): support openai cached prompt tokens in provider metadata

### PR DESCRIPTION
**Pull request by K-Mistele**
Adds support for OpenAI Prompt caching into `experimental_providerMetadata`. 
At @lgrammel's suggestion I tried to follow conventions in #3078

### CODE REVIEW

**Comment by K-Mistele** [comment]
cc @nicoalbanese

File: packages/openai/src/openai-chat-language-model.test.ts [hunk 1]
```diff
@@ -839,6 +842,36 @@ describe('doGenerate', () => {
     ]);
   });
 
+  it('should return cached_tokens in prompt_details_tokens', async () => {
+    // Difficult to test due to non-deterministic behavior of caching and eviction
+    // relates to when OpenAI has load
```
Comment (lgrammel) [inline packages/openai/src/openai-chat-language-model.test.ts:847]: please remove these 2 note lines

File: packages/openai/src/openai-chat-language-model.ts [hunk 2]
```diff
@@ -622,12 +440,207 @@ export class OpenAIChatLanguageModel implements LanguageModelV1 {
...
+          warnings,
+        };
+      }
+
+      default: {
+        const _exhaustiveCheck: never = type;
+        throw new Error(`Unsupported type: ${_exhaustiveCheck}`);
+      }
+    }
+  }
```
Comment (lgrammel) [inline packages/openai/src/openai-chat-language-model.ts:631]: please revert moving getArgs

File: packages/openai/src/openai-chat-language-model.ts [hunk 3]
```diff
@@ -622,12 +440,207 @@ export class OpenAIChatLanguageModel implements LanguageModelV1 {
...
 }
 
 const openAITokenUsageSchema = z
   .object({
     prompt_tokens: z.number().nullish(),
     completion_tokens: z.number().nullish(),
+    total_tokens: z.number().nullish(),
```
Comment (lgrammel) [inline packages/openai/src/openai-chat-language-model.ts:638]: total tokens not included on purpose, please remove

File: packages/openai/src/openai-chat-language-model.ts [hunk 4]
```diff
@@ -283,15 +95,21 @@ export class OpenAIChatLanguageModel implements LanguageModelV1 {
...
+      providerMetadata = { openai: {} };
+      if (response.usage?.completion_tokens_details?.reasoning_tokens) {
+        providerMetadata.openai.reasoningTokens =
+          response.usage?.completion_tokens_details?.reasoning_tokens;
+      }
+      if (response.usage?.prompt_tokens_details?.cached_tokens) {
+        providerMetadata.openai.cachedPromptTokens =
+          response.usage?.prompt_tokens_details?.cached_tokens;
+      }
+    }
```
Comment (lgrammel) [inline packages/openai/src/openai-chat-language-model.ts:112]: also needs support in doStream

Comment (lgrammel) [reply packages/openai/src/openai-chat-language-model.ts:112]: (reasoning tokens don't need doStream support because o1 does not support streaming)

Comment (K-Mistele) [reply packages/openai/src/openai-chat-language-model.ts:112]: ahhh got it.

File: packages/openai/src/openai-chat-language-model.ts (hunk 3 above)
Comment (K-Mistele) [reply packages/openai/src/openai-chat-language-model.ts:638]: understood.

File: packages/openai/src/openai-chat-language-model.ts (hunk 2 above)
Comment (K-Mistele) [reply packages/openai/src/openai-chat-language-model.ts]: done

File: packages/openai/src/openai-chat-language-model.test.ts (hunk 1 above)
Comment (K-Mistele) [reply packages/openai/src/openai-chat-language-model.test.ts]: done

File: packages/openai/src/openai-chat-language-model.ts (hunk 3 above)
Comment (K-Mistele) [reply packages/openai/src/openai-chat-language-model.ts]: done

File: packages/openai/src/openai-chat-language-model.ts (hunk 4 above)
Comment (K-Mistele) [reply packages/openai/src/openai-chat-language-model.ts:301]: not sure where to look for this - can't find anything about provider metadata in `doStream` for non-reasoning-related models. will dig in tomorrow.

Comment (K-Mistele) [reply packages/openai/src/openai-chat-language-model.ts]: added streaming support, example, and test @lgrammel :)

**Comment by K-Mistele** [comment]
per your message @lgrammel I updated & fixed tests; running `pnpm test` in `packages/openai` passes for me now.

File: packages/openai/src/openai-chat-language-model.ts [hunk 5]
```diff
@@ -613,6 +629,7 @@ export class OpenAIChatLanguageModel implements LanguageModelV1 {
                 promptTokens: usage.promptTokens ?? NaN,
                 completionTokens: usage.completionTokens ?? NaN,
               },
+              ...(providerMetadata != null ? { providerMetadata } : {}),
```
Comment (lgrammel) [inline packages/openai/src/openai-chat-language-model.ts:632]: You can just pass in `providerMetadata`, see Anthropic: https://github.com/vercel/ai/blob/main/packages/anthropic/src/anthropic-messages-language-model.ts#L419

**Review by lgrammel — APPROVED** [review]

**Review by lgrammel — CHANGES_REQUESTED** [review]
Looks great. Only the changeset is missing.

You can add a changeset by running `pnpm changeset` in the project root. Please make it a patch for the openai provider package.

File: packages/openai/src/openai-chat-language-model.ts (hunk 5 above)
Comment (K-Mistele) [reply packages/openai/src/openai-chat-language-model.ts]: Hmm this actually causes tests to fail due to a `toStrictEqual` assertion - it makes the providerMetadata present when it shouldn't be.

**Review by lgrammel — APPROVED** [review]

**Comment by K-Mistele** [comment]
pushed a commit to fix the prettier CI failure; seems related to the merge.

**Pull request merged**